usr/sbin/pkcsconf/pkcsconf.c

```diff
@@ -260,136 +230,79 @@ int main(int argc, char *argv[])
-            do {
-                printf("Enter the SO PIN: ");
-                fflush(stdout);
-                rc = get_pin(&sopin);
-            } while (rc == -EINVAL);
-        }
-        if (~flags & CFG_NEW_PIN) {
-            int rc;
-
-            do {
-                printf("Enter the new user PIN: ");
-                fflush(stdout);
-                rc = get_pin(&newpin);
-            } while (rc == -EINVAL);
-            newpinlen = strlen((char *) newpin);
-            do {
-                printf("Re-enter the new user PIN: ");
-                fflush(stdout);
-                rc = get_pin(&newpin2);
-            } while (rc == -EINVAL);
-            newpin2len = strlen((char *) newpin2);
-            if (newpinlen != newpin2len
-                || memcmp(newpin, newpin2, strlen((char *) newpin)) != 0) {
-                warnx("New PINs do not match.");
-                exit(CKR_PIN_INVALID);
-            }
+        if (!pin_so)
+            pin_so = pin_prompt(&buf_so, pin_prompt_so);
+
+        if (!pin_new)
+            pin_new = pin_prompt_new(&buf_new,
+                                     "Enter new USER PIN: ",
+                                     "Re-enter new USER PIN: ");
+
+        if (!pin_new) {
+            warnx("Invalid new user pin");
+            rv = CKR_PIN_INVALID;
+            goto done;
         }
-        rv = init_user_pin(in_slot, newpin, sopin);
+
+        rv = init_user_pin(in_slot, pin_new, pin_so);
+
+        /* partial cleanup/re-init for chained sub-functions */
+        pin_new = NULL;
+        pin_free(&buf_new);
     }
 
     /* If the user wants to set the SO PIN, check to see if they have passed the
      * current SO PIN and the New PIN in.  If not prompt and validate them. */
     if (flags & CFG_SET_SO) {
-        if (~flags & CFG_SO_PIN) {
-            int rc;
-
-            do {
-                printf("Enter the SO PIN: ");
-                fflush(stdout);
-                rc = get_pin(&sopin);
-            } while (rc == -EINVAL);
-        }
-        if (~flags & CFG_NEW_PIN) {
-            int rc;
-
-            do {
-                printf("Enter the new SO PIN: ");
-                fflush(stdout);
-                rc = get_pin(&newpin);
-            } while (rc == -EINVAL);
-            newpinlen = strlen((char *) newpin);
-            do {
-                printf("Re-enter the new SO PIN: ");
-                fflush(stdout);
-                rc = get_pin(&newpin2);
-            } while (rc == -EINVAL);
-            newpin2len = strlen((char *) newpin2);
-            if (newpinlen != newpin2len
-                || memcmp(newpin, newpin2, strlen((char *) newpin)) != 0) {
-                warnx("New PINs do not match.");
-                exit(CKR_PIN_INVALID);
-            }
+        if (!pin_so)
+            pin_so = pin_prompt(&buf_so, pin_prompt_so);
+
+        if (!pin_new)
+            pin_new = pin_prompt_new(&buf_new,
+                                     "Enter new SO PIN: ",
```

Comment:
Chang to `"Enter the new SO PIN: "` and `"Re-enter the new SO PIN: "`

Reply:
also check the other place to output the exact same message string as before in all cases.

Reply:
... also for the other tools. Changing the message text may cause regressions in existing scripts, also for distro partners, users, etc.